### ISSUE TITLE
Remove iteration in Enumerable#size

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#10014]
`Enumerable#size` is a confusing method. It looks inconspicuous, just return the size of a collection, no big deal. But it's actually pretty dangerous as its implementation `count { true }` can lead to infinite recursion on open enumerables. Since this is a pretty nasty thing, we should try to avoid that.

If an enumerable is finite there's often an efficient way to determine the size. Such implementing types should provide an optimized `size` implementation anyways. When there's no efficient way, they can fall back to `count { true }`.
If an enumerable is infinite, determining its size is impossible. So it doesn't make sense to count a `size` at all.

If you want to count elements in an (possibly) infinite enumerable type, you can always use `count { true }` explicitly, but then it's obvious that it iterates.

When `size` no longer works on inifinite enumerables, it can also serve as an indicator for other `Enumerable` methods that might run into infinite recursion to abort early. They would still need to *call* `size` because the pure existance of the method doesn't tell whether it returns a value. `Range#size` for example raises on open ended ranges.

It's probably best to not remove `Enumerable#size` entirely but change the implementation to raise at runtime. That way you can call `size` on unions collapsed to `Enumerable`.

As it turns out stdlib isn't much affected by this. `XML::Attributes` and `Range` (for non-int types) needed a `count { true }` implementation and that's it.
So I'm pretty confident that this breaking change won't be a huge disruption. Most including types already have a custom implementation (every `Indexable` for example).

**Comment by Blacksmoke16** [comment]
For `Range` you could probably also check if it's an endless range first to know if you can just return infinity.  Otherwise you would know its a finite range.

**Comment by straight-shoota** [comment]
That's already the case. `(1..).size` raises.

**Comment by asterite** [comment]
> If an enumerable is infinite, determining its size is impossible.

Also, map, select, count, etc., will also loop forever. Should we remove those too?

I don't understand why we should remove size only because an Enumerable can be infinite, when such case is pretty rare. If it's infinite, why would you call size on it?

**Comment by straight-shoota** [comment]
How do you know whether a particular `Enumerable` instance is infinite? There's no way to tell from the actual type. Some types are always finite, some always infinite but some (like `Range`) depend on the values.
Infinite enumerables might be rare, but it's just so easy to shoot yourself in the foot with seemingly trivial methods like `#size` causing your program to break.

But if `size` can be used to determine whether an `Enumerable` is infinite, we can use that in other methods as well, yes.

Yielding methods like `#map` should always work even with infinite enumerables because you can break out of the iteration at the call site. But methods that expect to consume the entire enumerable like `#to_a` should better raise when it's obvious that they can never complete.

**Comment by asterite** [comment]
@straight-shoota Makes sense.

I guess it's fine because in any case `size` is already defined in all stdlib types that include `Enumerable`.

**Comment by asterite** [comment]
That said, Ruby's Enumerable has `size` too, only it's called `count`: https://ruby-doc.org/core-2.7.2/Enumerable.html#method-i-count . That breaks with an infinite Enumerable. Why do they have it then?

**Comment by straight-shoota** [comment]
`#count` without arguments is IMO acceptable because the method name expresses to perform an action ("counting") whereas `#size` looks like only accessing a property.

**Comment by asterite** [comment]
Sounds good.

**Comment by straight-shoota** [comment]
> `size` is already defined in all stdlib types that include `Enumerable`.

That's not so much true, actually. There are just no test for them.
I've made `Enumerable#size` abstract to see where it's missing, and that's quite a few types. It's easy to fix. 

Maybe making it abstract wouldn't be the worst idea. Then including types are forced to decide what to do.

And that might be a good thing. `Dir#size` for example wouldn't work as you might expect when items have been read (related to #8504). And many enumerables wrapping internal collections (like `HTTP::Params`) wouldn't delegate `size`, so it would inefficiently iterate the items.

**Comment by straight-shoota** [comment]
That's the basic diff for making `Enumerable#size` abstract: https://github.com/straight-shoota/crystal/commit/c5cd694f492fec1ae441d7eeea983842e31e4516

Some of the NotImplementedError can probably be turned into proper implementations.

**Comment by asterite** [comment]
I don't understand why it needs to raise instead of failing to compile.

For Dir, with that reasoning, shouldn't to_a, map, etc., always operate on a newly opened dir?

**Comment by straight-shoota** [comment]
If failing on compile works, that's great. I'm just worried about enumerable types in a union being merged and then you can't call `size` even if all actual types implement it.

Dir is a different story in #8504

**Comment by oprypin** [comment]
I think it's good to not be able to pass any Enumerable into something that calls `.size`. With that expectation also comes the expectation that one can consume this number of items from the object, potentially multiple times. Which I don't think is always true (see `Dir`, `Char::Reader` -- though it's arguable whether it's just a failing of these particular classes)

**Comment by HertzDevil** [comment]
`Enumerator#size` does exist on Ruby, actually: https://ruby-doc.org/core-2.7.2/Enumerator.html#method-i-size

It returns an integer if the size is lazily known to be finite, `Float::INFINITY` if it's known to be infinite, and `nil` if the size cannot be determined lazily. More importantly, since their `Enumerator::Lazy < Enumerator`, that roughly corresponds to our `Iterator` which would all have lazily computed `#size`s.

`#count` is allowed to delegate to `#size` and some methods might benefit from knowing the aize in advance (e.g. `Array` capacity) but other than these I don't know how Ruby itself uses `Enumerator#size`.

**Comment by HertzDevil** [comment]
What I came up with in an hour:

```crystal
module Enumerable(T)
  # The class that signals that the collection has an infinite number of
  # elements.
  class Infinite
    INSTANCE = new
  end

  # Counts the number of elements in the collection by iterating over `self`.
  def count
    count { true }
  end

  # Returns the number of elements in the collection.
  #
  # Tries to determine the number of elements using `#length` first; if that
  # fails, calls `#count` instead.
  #
  # Raises `ArgumentError` if the collection is infinite.
  def size
    case length = self.length
    when Infinite then ArgumentError.new("Can't calculate size of an infinite Enumerable")
    when Nil      then count
    else               length
    end
  end

  # Attempts to determine the number of elements in the collection, without
  # iterating over `self`. This method should return `nil` if the number of
  # elements cannot be lazily computed, or `Enumerable::Infinite::INSTANCE` if
  # the collection is infinite.
  #
  # By default, this method returns `nil`, but including types may override. In
  # particular, `Indexable` requires its including types to always return a
  # finite length.
  def length
  end
end

module Indexable(T)
  # no need to make #size abstract (due to #9998 it already didn't work)

  # requires #9998 for this to work properly
  abstract def length : Int
end

# now including types should redefine #length instead of #size, e.g.

class Range(B, E)
  def length
    return Enumerable::Infinite::INSTANCE if @begin.nil? || @end.nil?
    # ...
  end
end

class StaticArray(T, N)
  def length : Int
    N
  end
end
```

`#length` sounds like a good candidate since we made sure to exclude most aliases from Ruby and now it is unused in the standard library (even `.length`). A second choice would be `#size?`.

IMO signalling infinite collections with a type is a better idea than `Float64::INFINITY` because none of the other floating-point sizes would ever make sense.

**Comment by asterite** [comment]
The reason we only have `size` is to avoid confusion between size, length, count, etc.

I personally think the existing approach (what's is master right now) is fine. What's the size of an Enumerable? You just traverse all elements to count them. Includers of Enumerable can redefine it to compute that more efficiently, but that's it. I really don't see a problem.

**Comment by straight-shoota** [comment]
> Includers of Enumerable can redefine it to compute that more efficiently

I don't agree that's a good idea.

a) It's not only about efficiency but primarily stability. Trying to consume an infinte enumerable entirely should raise as early as possible.
b) Includers even in stdlib tend to not redefine it, despite the default implementation being much less efficient and sometimes incorrect. See https://github.com/straight-shoota/crystal/commit/c5cd694f492fec1ae441d7eeea983842e31e4516

@HertzDevil I think your proposal is too much overengineered. There is no need for having more complexity than necessary. Removing the implementation of `Enumerable#size` or making it raise should be completely sufficient. We don't need a value to signal infinity, raising is just fine.

**Comment by asterite** [comment]
Can we override it where it's missing, and raise when we can detect it? I don't think we need to remove it from Enumerable...

**Comment by straight-shoota** [comment]
Ruby's `Enumerable` doesn't have `#size` either, only `#count` (and that's fine as it implies iteration).

**Comment by asterite** [comment]
Yes, but I'd like to avoid similarly named functions that do different things.

**Comment by straight-shoota** [comment]
That's a valid concern, but I don't think it weighs more than having a surprising iteration on `#size`. Also we don't need to add `#count` without argument/block just to replace `#size`. I'm fine with not having that as a convenience method. If you really want to count the items in an enumerable that doesn't have a fixed size, you would just use `count { true }` which makes it even more obvious that it iterates the collection.

**Comment by asterite** [comment]
Let me rephrase the first paragraph of this issue, but I'll use `map` instead of `size:

> `Enumerable#map` is a confusing method. It looks inconspicuous, just map collection, no big deal. But it's actually pretty dangerous as its implementation can lead to infinite recursion on open enumerables. Since this is a pretty nasty thing, we should try to avoid that.

So at least that paragraph isn't a reason to remove it. 

Let's continue.

> If an enumerable is finite there's often an efficient way to determine the size. Such implementing types should provide an optimized size implementation anyways. When there's no efficient way, they can fall back to count { true }.

Sounds reasonable to me! And that's how things were before #10162 was merged. So why did we change it?

> If an enumerable is infinite, determining its size is impossible. So it doesn't make sense to count a size at all.

Right, so don't call `size` or `count` on an inifinite enumerable. I never had an issue with this. When I know I'm dealing with an infinite enumerator I just don't call `size` on it. Infinite enumerators usually come from `cycle`, so when doing `cycle.size` it's pretty clear that's not going to finish. As clear as doing `cycle.count`.

**Comment by asterite** [comment]
I think, given that we are very close to 1.0, the best thing to do is to revert #10162 and to continue discussing all of this after 1.0. If this isn't going to be a breaking change then we can postpone the discussion.

**Comment by straight-shoota** [comment]
"It looks inconspicuous" does not apply in the same way. `map` clearly needs to iterate every item in the collection no matter what. `size` looks like a simple getter and for most implementing types it is just that which just emphasizes that interpretation. But when you end up with the default implementation of `Enumerable#size`, it isn't as simple as that.

When I know I'm dealing with an infinite enumerator I just don't call `size` on it. How do you know whether a collection is infinite? Or how do you know if calling `size` actually consumes it?

**Comment by asterite** [comment]
> But how do you know whether a collection is infinite?

You know when it hangs, or when it doesn't.

In Ruby you can call `count` on an Enumerable. But how do you know whether a collection is infinite? Isn't this the same problem? I don't understand.

I think the whole issue is "count is a more descriptive English word that conveys the meaning of traversal, while size isn't, so let's have both words do different things". Right?

**Comment by straight-shoota** [comment]
It's the same problem but in Ruby there's a solution missing in Crystal: `Enumerator` class has `size` as an alternative which doesn't hang.

> I think the whole issue is "count is a more descriptive English word that conveys the meaning of traversal, while size isn't, so let's have both words do different things". Right?

Yeah, kind of. If you want to know how many elements are in a collection, some have the amount easily accessible (`size`), others need to iterate to know what (`count`). And that's very different methods of operation.

**Comment by asterite** [comment]
> And that's very different methods of operation.

Yes, they have different implementation details. But in the end the do the answer the same question: how many elements are there in this enumerable?

In general people don't care about implementation details. Ask a Ruby developer, would they use "size" or "count" when dealing with a collection, I'm sure nobody will know the difference between those.

I'm still waiting for a use case for Ruby's `size` method.

**Comment by asterite** [comment]
I'll wait for others to comment on this, particularly @bcardiff and @waj . Maybe adding `count` isn't that bad after all. But we need to add it to all existing types (they are not many, and most include `Indexable` already).

If we go this route, I'd also like to add `take(n)` to `Enumerable` and `Iterator` as an almost alias of `first(n)`.
